File: 13_FibonacciNumbers_FibFrog.rb
# you can write to stdout for debugging purposes, e.g.
# puts "this is a debug message"

def solution(a)
    # write your code in Ruby 2.2
    # n + 1 以下のフィボナッチ数列を列挙
    # n + 1 がフィボナッチ数列に含まれているならば回数1で到達できる
    n = a.length
    fibo = []
    fibo << 1 
    fibo << 2
    while fibo[-1] + fibo[-2] <= n + 1
      fibo << fibo[-1] + fibo[-2]
    end
  
    big = 10 ** 10
    leaf = [1]
    leaf << a 
    leaf << [1]
    leaf.flatten! # [1, a1, a2, ... , an-1, an, 1] の形(両岸を1にする)
    n = leaf.length
    arr = Array.new(n, big) # 距離iに到達できる最小回数を記録
    arr[0] = 0
    0.upto(n - 1) do |i|
      next if arr[i] == big
      fibo.each do |f|
        dist = i + f
        #p dist
        next if dist >= n || leaf[dist] == 0
        arr[dist] = arr[i] + 1 if arr[dist] > arr[i] + 1
      end
    end
    
    # p n
    # p fibo 
    # p leaf
    # p arr
    return arr[-1] != big ? arr[-1] : -1
  end
  

File: Python/13_FibonacciNumbers_FibFrog.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A):
    # write your code in Python 3.6
		# 向こう岸も葉っぱとして考えると楽
    A += [1]
    n = len(A)  

    # とりあえずフィボナッチを列挙する
    fibo = [1, 2]
    f = 3 
    while f <= n:
        fibo.append(f)
        f += fibo[-2]

    # 先頭から順にfiboで飛べるか見ていく
    # O(N * len(fibo))で，n=10^6のときでもlen(fibo)=29なので大丈夫
    flag = [0] * (n + 1)
    flag[0] = 1	# 最初の岸も葉っぱとして考えると楽
    for i in range(n + 1):
        # 葉っぱでない or 葉っぱだけど到達できないなら考えない
        if flag[i] == 0: continue   

        for f in fibo:
            if f + i >= n + 1 or A[f + i - 1] == 0: continue   

						# a番目の葉っぱに到達できる最小ジャンプ回数を記録していく
            if flag[f + i] == 0: flag[f + i] = flag[i] + 1
            else: flag[f + i] = min(flag[f + i], flag[i] + 1)
    
    return flag[-1] - 1 if flag[-1] > 0 else -1
# end of func
    
